### PR TITLE
High robot

### PR DESCRIPTION
**Pull request by EganJ**
Has some untested features, but is the most advanced branch: other things have started to branch off of this, so we should merge to main to avoid confusion.

### CODE REVIEW

File: .wpilib/wpilib_preferences.json [hunk 1]
```diff
@@ -2,5 +2,5 @@
     "currentLanguage": "java",
     "enableCppIntellisense": false,
     "projectYear": "2023",
-    "teamNumber": 1745
+    "teamNumber": 8874
```
Comment (edenbrown828) [inline .wpilib/wpilib_preferences.json:5]: Are we going to have to keep changing this back and forth or is there a way we can avoid that?

File: src/main/deploy/pathplanner/Score .path [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      "stopEvent": {
+        "names": [],
+        "executionBehavior": "parallel",
+        "waitBehavior": "none",
+        "waitTime": 0
+      }
+    }
+  ],
+  "markers": []
+}
```
Comment (edenbrown828) [inline src/main/deploy/pathplanner/Score .path:49]: Where is the documentation explaining the definition/significance of each of these values, and if there isn't one central place where everything is written down can we create one?

File: src/main/java/frc/robot/Constants.java [hunk 3]
```diff
@@ -152,8 +154,9 @@ public static final class ModuleConstants {
     public static final double kDrivingMinOutput = -1.0;
     public static final double kDrivingMaxOutput = 1.0;
 
+    //was 3.75,0,.35
     public static final PIDGains kTurningPIDGains =
-        new PIDGains(4, 0, 0.075); // TODO: tune values for the turning motor
+        new PIDGains(6.75, 0.03, 0.35); // TODO: tune values for the turning motor
```
Comment (nbgray) [inline src/main/java/frc/robot/Constants.java:159]: Are these tuned values good now?  Can we remove the TODO?

File: src/main/java/frc/robot/Constants.java [hunk 4]
```diff
@@ -174,68 +177,109 @@ public static final class AutoConstants {
...
-    public static final double kPThetaController = 1.0;
-
+    public static final double kMaxAngularSpeedRadiansPerSecondSquared = Math.PI;    
+    
     // Constraint for the motion profiled robot angle controller
     public static final TrapezoidProfile.Constraints kThetaControllerConstraints =
         new TrapezoidProfile.Constraints(
             kMaxAngularSpeedRadiansPerSecond, kMaxAngularSpeedRadiansPerSecondSquared);
+
+    // public static final PIDController kPXController = new PIDController(0.0, 0, 0.0);// 1.0 - 0.35
```
Comment (nbgray) [inline src/main/java/frc/robot/Constants.java]: Please delete commented out code

File: src/main/java/frc/robot/Constants.java [hunk 5]
```diff
@@ -174,68 +177,109 @@ public static final class AutoConstants {
...
-    public final static double INTAKE_ARM_SHELF = 0.0;
-    public final static double SCORING_ARM_LOW = 0.0;
-    public final static double SCORING_ARM_MID = 0.0;
-    public static final Double INTAKE_ARM_STOWED = 0.0;
+    public final static float ARM_REVERSE_LIMIT = 34.0f;
+    public final static float ARM_FORWARD_LIMIT = 0.0f;
+    public final static double ARM_REVERSE_SPEED = 0.5;
+    public final static double ARM_FORWARD_SPEED = 0.5;
+
+    // Elbow limits // TODO: tune these values
```
Comment (nbgray) [inline src/main/java/frc/robot/Constants.java:210]: Are these values tuned?  Can we delete the TODO?

File: src/main/java/frc/robot/Constants.java [hunk 6]
```diff
@@ -174,68 +177,109 @@ public static final class AutoConstants {
...
+    // Idle mode
+    public static final IdleMode kMotorIdleMode = IdleMode.kBrake; 
+
+    // Current limits 
     public static final int kArmMotorCurrentLimit = 40;
-    public static final int kMovingArmMotorCurrentLimit = 60;  
+    public static final int kMovingArmMotorCurrentLimit = 60;
+    public static final int kElbowMotorCurrentLimit = 60;
 
     // PID constants using custom PID gains class //TODO: tune PID values
```
Comment (nbgray) [inline src/main/java/frc/robot/Constants.java:242]: Are these PID constants tuned?  Can we delete the TODO?

File: src/main/java/frc/robot/RobotContainer.java [hunk 7]
```diff
@@ -15,115 +14,173 @@
...
-  /** The container for the robot. Contains subsystems, OI devices, and commands. */
-  public RobotContainer(Localizer localizer) {
-     
-    m_localizer = localizer;
-   
-   
     // Configure the trigger bindings
-    configureBindings();  
+    configureDriverBindings();
+    configureOpperatorBindings();
```
Comment (nbgray) [inline src/main/java/frc/robot/RobotContainer.java]: Operator spelling

File: src/main/java/frc/robot/RobotContainer.java [hunk 8]
```diff
@@ -15,115 +14,173 @@
...
+    new JoystickButton(m_driverController, XboxController.Button.kX.value).onTrue(armPositions.armPickUpCubeCommand())
+        .whileTrue(new RunCommand(() -> intake.manipulates(1.0))).onFalse(armPositions.armStowCommand())
+        .whileFalse(new RunCommand(() -> intake.manipulates(direction)));
+    // Shelf
+    new JoystickButton(m_operatorController, XboxController.Button.kLeftBumper.value)
+        .onTrue(armPositions.armPickUpFromShelf()).whileTrue(new RunCommand(() -> intake.manipulates(-1.0)));
+  }
 
-          //Make sure that all buttons are unique
+  private void configureOpperatorBindings() {
```
Comment (nbgray) [inline src/main/java/frc/robot/RobotContainer.java]: Operator spelling

File: src/main/java/frc/robot/auton/util/AutonBase.java [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+    AutoChooser.addAuton(this, name);
+  }
+
+  public AutonBase(String name) {
+    AutoChooser.addAuton(this, name);
+  }
+
+
+// TODO: Possibly look into a low priority issue for changing printing to logging
```
Comment (nbgray) [inline src/main/java/frc/robot/auton/util/AutonBase.java:23]: Is this TODO resolved?

File: src/main/java/frc/robot/auton/util/AutonBase.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+// TODO: Possibly look into a low priority issue for changing printing to logging 
+  public void addCommandsWithLog(String tag, Command... commands) {
+    // Add Printing to each command
+
+    // Print start of group
+    super.addCommands(
+        new InstantCommand(
+            () -> {
+              m_autonStartTime = Timer.getFPGATimestamp();
+              System.out.println("Auton group Starting at {}" + m_autonStartTime);
```
Comment (nbgray) [inline src/main/java/frc/robot/auton/util/AutonBase.java:32]: If you want the time in the brackets, you'll need to change this to '"Auton group Starting at {" + m_autonStartTime + "}"'

And the same for all the future logs too.

File: src/main/java/frc/robot/auton/util/AutonUtil.java [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+        return new Pose2d(
+            trajectory.getInitialState().poseMeters.getTranslation(),
+            trajectory.getInitialState().holonomicRotation);
+      }
+    
+      public static PathPlannerTrajectory loadTrajectory(String name, double maxVel, double maxAccel) {
+        return loadTrajectory(name, maxVel, maxAccel, false);
+      }
+
+      // public static PathPlannerTrajectory load
```
Comment (nbgray) [inline src/main/java/frc/robot/auton/util/AutonUtil.java]: Delete commented out code

File: src/main/java/frc/robot/commands/armCommands/PositionCommand.java [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+    ShoulderPosition shoulderPosition;
+    Arm arm;
+    Intake intake;
+
+    public PositionCommand(Arm armMech, Intake intakeMech){
+        this.arm = armMech;
+        this.intake = intakeMech;
+    }
+    
+    // public SequentialCommandGroup armStowCommand(){
```
Comment (nbgray) [inline src/main/java/frc/robot/commands/armCommands/PositionCommand.java]: Should all these commented out sections be uncommented?  Or deleted?

File: src/main/java/frc/robot/commands/armCommands/ShoulderPosition.java [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+        this.position = position;
+        addRequirements(arm);
+    }
+
+    @Override
+    public void execute() {
+        arm.setArmPosition(position);
+    }
+
+    // TODO: Make this better and more accurate
```
Comment (nbgray) [inline src/main/java/frc/robot/commands/armCommands/ShoulderPosition.java:21]: Is this TODO resolved?

File: src/main/java/frc/robot/subsystems/arm/ArmPositions.java [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+        public static double ARM_PICK_UP_SHELF = 0.445;
+        public static double ELBOW_PICK_UP_SHELF = 0.83;
+        // Arm back cone pick up scoring position
+        public static double ARM_BACK_PICKUP = 0.0;
+        public static double ELBOW_BACK_PICKUP = 0.0;
+        public static double ARM_SPECIAL_IDLE = 0.0;
+        public static double ELBOW_SPECIAL_IDLE = 0.0;
+        // Elbow Idle position
+        public static double ELBOW_IDLE = .83;
+        // public static double ARM_STOWED = .55;
```
Comment (nbgray) [inline src/main/java/frc/robot/subsystems/arm/ArmPositions.java]: Should all these commented out constants be deleted?

File: vendordeps/NavX.json [hunk 15]
```diff
@@ -0,0 +1,39 @@
+{
```
Comment (nbgray) [inline vendordeps/NavX.json:1]: I thought we got rid of the NavX?  Shouldn't this file be deleted?

**Review by nbgray — CHANGES_REQUESTED** [review]

File: src/main/java/frc/lib/swerve/MAXSwerve.java [hunk 16]
```diff
@@ -297,6 +297,10 @@ public double getHeading() {
     // return m_gyro.getRotation2d().getDegrees();
   }
 
+  public double getAdjustedAngle(){
```
Comment (edenbrown828) [inline src/main/java/frc/lib/swerve/MAXSwerve.java:300]: Add comments explaining this method and the ones below it

File: src/main/java/frc/robot/Constants.java [hunk 17]
```diff
@@ -174,68 +177,109 @@ public static final class AutoConstants {
...
+    
     // Constraint for the motion profiled robot angle controller
     public static final TrapezoidProfile.Constraints kThetaControllerConstraints =
         new TrapezoidProfile.Constraints(
             kMaxAngularSpeedRadiansPerSecond, kMaxAngularSpeedRadiansPerSecondSquared);
+
+    // public static final PIDController kPXController = new PIDController(0.0, 0, 0.0);// 1.0 - 0.35
+    // public static final PIDController kPYController = new PIDController(0.0, 0, 0.0);// 1.075 - 0.45
+    // public static final PIDController kPThetaController = new PIDController(0.0, 0, 0.0);// 6.0 - 0.5
+    public static final PIDController kPXController = new PIDController(1.0, 0, 0.35);// 1.0 - 0.35
```
Comment (edenbrown828) [inline src/main/java/frc/robot/Constants.java:188]: What do the numbers in the comments mean?  Can we delete them?

File: src/main/java/frc/robot/auton/locations/BottomStart.java [hunk 18]
```diff
@@ -0,0 +1,16 @@
...
+public class BottomStart extends Locations{
+    private BottomStart(){}
+
+    @Override
+    public Pose2d get() {
+        return new Pose2d(1.81, 0.44, Rotation2d.fromDegrees(180));
+    }
+
+    public static final BottomStart instance = new BottomStart();
+
```
Comment (edenbrown828) [inline src/main/java/frc/robot/auton/locations/BottomStart.java:15]: Add comments throughout this file please

File: src/main/java/frc/robot/auton/locations/Locations.java [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+  }
+
+public static BottomStart BottomStart(){
+    return BottomStart.instance;
+    }
+
+public static MiddleStart MiddleStart(){
+    return MiddleStart.instance;
+    }
+
```
Comment (edenbrown828) [inline src/main/java/frc/robot/auton/locations/Locations.java:24]: Add comments throughout this file please

File: src/main/java/frc/robot/auton/locations/MiddleStart.java [hunk 20]
```diff
@@ -0,0 +1,16 @@
...
+public class MiddleStart extends Locations {
+    private MiddleStart() {}
+
+    @Override
+    public Pose2d get() {
+        return new Pose2d(1.81, 2.21, Rotation2d.fromDegrees(180));
+    }
+
+    public static final MiddleStart instance = new MiddleStart();
+
```
Comment (edenbrown828) [inline src/main/java/frc/robot/auton/locations/MiddleStart.java:15]: Add comments throughout this file please

File: src/main/java/frc/robot/auton/locations/TopStart.java [hunk 21]
```diff
@@ -0,0 +1,16 @@
...
+public class TopStart extends Locations{
+    private TopStart(){}
+
+    @Override
+    public Pose2d get() {
+        return new Pose2d(1.81, 4.94, Rotation2d.fromDegrees(180));
+    }
+
+    public static final TopStart instance = new TopStart();
+
```
Comment (edenbrown828) [inline src/main/java/frc/robot/auton/locations/TopStart.java:15]: Add comments throughout this file please

File: src/main/java/frc/robot/auton/paths/top/TopPark.java [hunk 22]
```diff
@@ -0,0 +1,40 @@
...
+        Constants.AutoConstants.kPYController,
+        Constants.AutoConstants.kPThetaController),
+    new InstantCommand(() -> drive.drive(0.0, 0.0, 0.0, false), drive));
+
+
+    }
+    @Override
+  public void initSendable(SendableBuilder builder) {
+    super.initSendable(builder);
+  }
```
Comment (edenbrown828) [inline src/main/java/frc/robot/auton/paths/top/TopPark.java:39]: Add comments throughout this file please

File: src/main/java/frc/robot/auton/util/AutonUtil.java [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
+         * Store trajectories. This is done since multiple autons may load the same trajectories,
+         * no need to load it again and again. If anything funky with trajectories is happening,
+         * comment out the below. (e.g. I don't see any stored state in the trajectory, but it
+         * is not clear if that is true, and if calling the trajectory stores any state.
+         * Investigate later.)
+         */
+        m_cache.put(cacheKey_, trajectory);
+    
+        return trajectory;
+      }
```
Comment (edenbrown828) [inline src/main/java/frc/robot/auton/util/AutonUtil.java:59]: Add more comments throughout this file please

File: src/main/java/frc/robot/commands/armCommands/ElbowPosition.java [hunk 24]
```diff
@@ -0,0 +1,26 @@
...
+        this.position = position;
+        addRequirements(arm);
+    }
+    
+    @Override
+    public void execute() {
+        arm.setElbowPosition(position);
+    }
+
+    // TODO: Make this better and more accurate
```
Comment (edenbrown828) [inline src/main/java/frc/robot/commands/armCommands/ElbowPosition.java:21]: Is this todo resolved or can it be resolved?

File: src/main/java/frc/robot/subsystems/arm/Arm.java [hunk 25]
```diff
@@ -4,115 +4,182 @@
...
-    // braking)
-    motor.enableSoftLimit(CANSparkMax.SoftLimitDirection.kForward, true);
-    motor.setSoftLimit(CANSparkMax.SoftLimitDirection.kForward, Constants.ArmConstants.FORWARD_LIMIT);
-    motor.enableSoftLimit(CANSparkMax.SoftLimitDirection.kReverse, true);
-    motor.setSoftLimit(CANSparkMax.SoftLimitDirection.kReverse,  Constants.ArmConstants.REVERSE_LIMIT);
-
+    // set motor basic values
+    motor.setIdleMode(Constants.ArmConstants.kMotorIdleMode);
+    // motor.enableSoftLimit(CANSparkMax.SoftLimitDirection.kForward, false);
+    // motor.enableSoftLimit(CANSparkMax.SoftLimitDirection.kReverse, false);
```
Comment (edenbrown828) [inline src/main/java/frc/robot/subsystems/arm/Arm.java]: delete or uncomment commented code

File: src/main/java/frc/robot/subsystems/arm/Arm.java [hunk 26]
```diff
@@ -4,115 +4,182 @@
...
-        Constants.ArmConstants.ARM_MIN_OUTPUT, Constants.ArmConstants.ARM_MAX_OUTPUT);
+        Constants.ArmConstants.MIN_OUTPUT, Constants.ArmConstants.MAX_OUTPUT);
+  //       motor.setPeriodicFramePeriod(PeriodicFrame.kStatus0, 20);
+  //       motor.setPeriodicFramePeriod(PeriodicFrame.kStatus5, 20);
+  // motor.setPeriodicFramePeriod(PeriodicFrame.kStatus1, 200);
+  // motor.setPeriodicFramePeriod(PeriodicFrame.kStatus2, 20);
+  // motor.setPeriodicFramePeriod(PeriodicFrame.kStatus3, 200);
+  // motor.setPeriodicFramePeriod(PeriodicFrame.kStatus4, 200);
+// motor.setSoftLimit(CANSparkMax.SoftLimitDirection.kForward, Constants.ArmConstants.ARM_FORWARD_LIMIT);
+//     motor.setSoftLimit(CANSparkMax.SoftLimitDirection.kReverse, Constants.ArmConstants.ARM_REVERSE_LIMIT);
```
Comment (edenbrown828) [inline src/main/java/frc/robot/subsystems/arm/Arm.java]: delete or uncomment commented code

File: src/main/java/frc/robot/subsystems/arm/Arm.java [hunk 27]
```diff
@@ -4,115 +4,182 @@
...
+  elbowPIDController.setOutputRange(
+      Constants.ArmConstants.MIN_OUTPUT, Constants.ArmConstants.MAX_OUTPUT);
+  //     motor.setPeriodicFramePeriod(PeriodicFrame.kStatus0, 20);
+  // motor.setPeriodicFramePeriod(PeriodicFrame.kStatus5, 20);
+  // motor.setPeriodicFramePeriod(PeriodicFrame.kStatus1, 200);
+  // motor.setPeriodicFramePeriod(PeriodicFrame.kStatus2, 20);
+  // motor.setPeriodicFramePeriod(PeriodicFrame.kStatus3, 200);
+  // motor.setPeriodicFramePeriod(PeriodicFrame.kStatus4, 200);
+      // motor.setSoftLimit(CANSparkMax.SoftLimitDirection.kForward, Constants.ArmConstants.ELBOW_FORWARD_LIMIT);
+      // motor.setSoftLimit(CANSparkMax.SoftLimitDirection.kReverse, Constants.ArmConstants.ELBOW_REVERSE_LIMIT);
```
Comment (edenbrown828) [inline src/main/java/frc/robot/subsystems/arm/Arm.java]: Delete or uncomment commented code

File: src/main/java/frc/robot/subsystems/arm/Arm.java [hunk 28]
```diff
@@ -4,115 +4,182 @@
...
-    armPositions.put(armPosition.SCORING_ARM_POSITION_MID, Constants.ArmConstants. SCORING_ARM_MID);
-    armPositions.put(armPosition.INTAKE_ARM_POSITION_STOWED,  Constants.ArmConstants.INTAKE_ARM_STOWED);
-  
-  
-    this.resetArmPosition();
+    // armPositions.put(armPosition.INTAKE_ARM_POSITION_GROUND,  Constants.ArmConstants.INTAKE_ARM_GROUND);
+    // armPositions.put(armPosition.INTAKE_ARM_POSITION_SHELF,  Constants.ArmConstants.INTAKE_ARM_SHELF);
+    // armPositions.put(armPosition.SCORING_ARM_POSITION_LOW,  Constants.ArmConstants.SCORING_ARM_LOW);
+    // armPositions.put(armPosition.SCORING_ARM_POSITION_MID, Constants.ArmConstants. SCORING_ARM_MID);
+    // armPositions.put(armPosition.INTAKE_ARM_POSITION_STOWED,  Constants.ArmConstants.INTAKE_ARM_STOWED);
```
Comment (edenbrown828) [inline src/main/java/frc/robot/subsystems/arm/Arm.java]: Delete or uncomment commented code

File: src/main/java/frc/robot/subsystems/arm/ArmPositions.java [hunk 29]
```diff
@@ -0,0 +1,174 @@
...
+        // Score Cone High
+        SmartDashboard.putNumber("Arm Cone High", Positions.ARM_SCORE_CONE_HIGH);
+        SmartDashboard.putNumber("Elbow Cone High", Positions.ELBOW_SCORE_CONE_HIGH);
+        // Elbow Idle Position
+        SmartDashboard.putNumber("Elbow Idle", Positions.ELBOW_IDLE);
+    }
+
+    public static class Positions {
+        // Arm Stowed Positions
+        public static double ARM_STOWED = .55;
```
Comment (edenbrown828) [inline src/main/java/frc/robot/subsystems/arm/ArmPositions.java:38]: These are named like constants - are they not?  Can they be made into constants or the casing of the names changed?

File: src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java [hunk 30]
```diff
@@ -51,11 +55,11 @@ public DriveSubsystem(ADIS16470_IMU m_gyro) {
 
 
   // TODO: Add sendable data for controllers during autonomous mode
```
Comment (edenbrown828) [inline src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:57]: Is this TODO resolved?

File: src/main/java/frc/robot/subsystems/intake/Intake.java [hunk 31]
```diff
@@ -1,113 +1,85 @@
...
-//   return new Rotation2d(getPosition() - Constants.Intake.kIntakeOffset);
-// }
-
-// sets the intake motor to 0 to stop all movement 
-public void idle() {
+public void stop(){
   intakeMotor.set(0.0);
 }
 
-public void MunaulPickup(){
```
Comment (edenbrown828) [inline src/main/java/frc/robot/subsystems/intake/Intake.java:90]: Misspelled manual

File: src/main/java/frc/robot/Constants.java (hunk 3 above)
Comment (QuantumOtterDev) [reply src/main/java/frc/robot/Constants.java:159]: No they need to be retuned with the feedforward

File: src/main/java/frc/robot/Constants.java (hunk 6 above)
Comment (QuantumOtterDev) [reply src/main/java/frc/robot/Constants.java:242]: They need to be retuned with the feedforward controller

File: .wpilib/wpilib_preferences.json (hunk 1 above)
Comment (QuantumOtterDev) [reply .wpilib/wpilib_preferences.json:5]: you can't currently due to how wpilib wraps their team number, unless you unwrap it and make custom code to handle that, Off season problem at this point

File: src/main/deploy/pathplanner/Score .path (hunk 2 above)
Comment (QuantumOtterDev) [reply src/main/deploy/pathplanner/Score .path:49]: This is all pathplanner built in json deployer, they have some documentation to explain how it works, that lead to this file being straight forward, at this point it's a off season thing

File: src/main/java/frc/robot/Constants.java (hunk 4 above)
Comment (at384752) [reply src/main/java/frc/robot/Constants.java]: deleted

File: src/main/java/frc/robot/RobotContainer.java (hunk 7 above)
Comment (at384752) [reply src/main/java/frc/robot/RobotContainer.java]: fixed

File: src/main/java/frc/robot/auton/util/AutonUtil.java (hunk 11 above)
Comment (at384752) [reply src/main/java/frc/robot/auton/util/AutonUtil.java]: deleted

File: src/main/java/frc/robot/commands/armCommands/PositionCommand.java (hunk 12 above)
Comment (at384752) [reply src/main/java/frc/robot/commands/armCommands/PositionCommand.java]: comments have been deleted

File: src/main/java/frc/robot/subsystems/arm/ArmPositions.java (hunk 14 above)
Comment (at384752) [reply src/main/java/frc/robot/subsystems/arm/ArmPositions.java]: comments have been deleted

File: src/main/java/frc/robot/Constants.java (hunk 17 above)
Comment (at384752) [reply src/main/java/frc/robot/Constants.java:188]: comments have been deleted

File: src/main/java/frc/robot/subsystems/arm/Arm.java (hunk 27 above)
Comment (at384752) [reply src/main/java/frc/robot/subsystems/arm/Arm.java]: commented code has been deleted

File: src/main/java/frc/robot/subsystems/arm/Arm.java (hunk 28 above)
Comment (at384752) [reply src/main/java/frc/robot/subsystems/arm/Arm.java]: deleted

File: src/main/java/frc/robot/subsystems/intake/Intake.java (hunk 31 above)
Comment (at384752) [reply src/main/java/frc/robot/subsystems/intake/Intake.java:90]: the code has been deleted by someone else

**Review by nbgray — APPROVED** [review]
Changes look good

**Pull request merged**